datafusion/expr/src/logical_plan/plan.rs

```diff
@@ -603,6 +604,73 @@ impl LogicalPlan {
         Ok(new_plan)
     }
 
+    /// Walk the logical plan, find any `PlaceHolder` tokens, and return a map of their IDs and DataTypes
+    pub fn get_parameter_types(
+        &self,
+    ) -> Result<HashMap<String, Option<DataType>>, DataFusionError> {
+        struct ParamTypeVisitor {
+            param_types: HashMap<String, Option<DataType>>,
+        }
+
+        struct ExprParamTypeVisitor {
+            param_types: HashMap<String, Option<DataType>>,
+        }
+
+        impl ExpressionVisitor for ExprParamTypeVisitor {
+            fn pre_visit(
+                mut self,
+                expr: &Expr,
+            ) -> datafusion_common::Result<Recursion<Self>>
+            where
+                Self: ExpressionVisitor,
+            {
+                if let Expr::Placeholder { id, data_type } = expr {
+                    let _ = self.param_types.insert(id.clone(), data_type.clone());
+                }
+                Ok(Recursion::Continue(self))
+            }
+        }
+
+        impl PlanVisitor for ParamTypeVisitor {
+            type Error = DataFusionError;
+
+            fn pre_visit(&mut self, plan: &LogicalPlan) -> Result<bool, Self::Error> {
+                let mut visitor = ExprParamTypeVisitor {
+                    param_types: Default::default(),
+                };
+                match plan {
```

Comment:
There are probably more plan types (like Aggregates for example) that could have references to parameter values

I poked around for a "visit all `Expr` in a plan" preexisting code and could not find one. So longer term I think it would be good to move this code somewhere more general (maybe near `ExpressionVisitor`):

```rust
/// visit all expressions in all nodes of the LogicalPlan
fn visit_plan_exprs(plan: &LogicalPlan, visitor: ExpressionVisitor) {
...
}

However think it is ok to start with this approach of something very specialized for parameter type inference and then refactor into something more general as a follow on PR. I also think we can track adding support for parameters in other plan types with a ticket as well if you don't want to support the entire thing in the first PR

Reply:
> if you don't want to support the entire thing in the first PR

I think that would be best.

Reply:
> a) Extract the "walk all exprs in a plan" code to its own function
> b) Cover the remaining LogicalPlan types in "walk all exprs in a plan"

I think these are done now.